nativescript-3.2.0-memory-management/README.md

```diff
@@ -0,0 +1,116 @@
+    - A splice is creates for "extended native class", such as `var MyView = View.extends({ ... }); var myView = new MyView();`
+ - it is considered **not to** "have implementation object" in cases when:
+    - A splice is created for `var button = new android.widget.Button(...)` where a native class is instantiated
+    - A splice is created for `var i = anAndroidObject.getValue()` when Java instance is returned by the getValue() method
+
+When Splice is created
+ - The Java instance is set strong, and can not be collected by the Dalvik GC
+
+On V8 GC collection phase:
+ - For each splice that has implementation object, from its JavaScript instance, all other reachable JavaScript instances are traversed. For these reachable JavaScript objects:
+    - If the reached object is an implementation object - the Java instance of its splice is set weak
+    - Otherwise it is ignored
+
+Then all splices may fall in one of these cases:
+ - If JavaScript instance is marked for collection and has no implementation object, then the JavaScript instance is let to be collected, and the Java instance is set weak
+ - If JavaScript instance is marked for collection and has implementation object and its Java instance is already set weak, then if the Java object is alive, the JavaScript instance is revived, else if the Java object is dead, the JavaScript instance is let to be collected
+ - If JavaScript instance is marked for collection and has implementation object, then the Java instance is set weak (this happens once the first time the instance is marked for collection), the JavaScript instance is revived
+
+### Properties of the Implementation
+#### Premature Collection
+Unlike iOS, both the Java and JavaScript in the Android runtime are managed. The native framework rarely relies on weak references, so premature collection can hardly be observed. Change in behavior after a GC for Android more frequently result in the "Half-Dead Splice".
+
+#### Leaks
+Memory leaks are rare. If there is a pool of unreachable splices from either Java or JavaScript, at some point the V8 GC will notify the JavaScript instances they are marked for collection, then the Java counterparts will be set weak. Then after Dalvik GC the Java instances will be collected. At the next V8 GC the JavaScript instances will be let for collection as the Java instances are dead.
+
+#### Half-Dead Splice
+Since collection is driven by the garbage collectors it is possible to hold weak reference to the JavaScript instance of a splice. After V8 GC it can set weak the Java instance. Dalvik GC can collect it. Then before the next V8 GC the JavaScript instance can be obtained from the weak ref and methods can be called that result in accessing half-dead splice since the Java counterpart is dead already. The errors reported points that we fail to find object for id. These problems are perceived as random and are quite hard to reproduce.
```

Comment:
`it can set weak the Java instance` it can set the Java instance to hold weak reference